### DISCUSSION TITLE
Add assemblyof operator to complement typeof and nameof

### DISCUSSION THREAD

**Discussion by jbogard** [dotnet/csharplang#2715] [Language Ideas]
It would still take a type (not a namespace):

```c#
var assembly = typeof(Foo).Assembly;
//or
var assembly = assemblyof(Foo);
```

**Comment by HaloFour** [comment]
Aside from saving 5 characters, what would the benefit of `assemblyof(T)` be over `typeof(T).Assembly`?  How often do you need to get an `Assembly` instance for a given type?

**Comment by jbogard** [comment]
Libraries/frameworks that perform some sort of assembly scanning are my main use case. These libraries accept `Assembly` instances as parameters. During a configuration step of an application, they'll scan those assemblies for various implemented extensions.

So any of our applications, that's 2, 3, 4 or more cases.

These libraries typically expose a couple ways of scanning:
 - Type
 - Assembly

A good example is the Scrutor project:

```c#
var collection = new ServiceCollection();

collection.Scan(scan => scan
    .FromAssemblyOf<ITransientService>()
        .AddClasses(classes => classes.AssignableTo<ITransientService>())
            .AsImplementedInterfaces()
            .WithTransientLifetime()
        .AddClasses(classes => classes.AssignableTo<IScopedService>())
            .As<IScopedService>()
            .WithScopedLifetime()
        .AddClasses(classes => classes.AssignableTo(typeof(IOpenGeneric<>)))
            .AsImplementedInterfaces()
        .AddClasses(classes => classes.AssignableTo(typeof(IQueryHandler<,>)))
            .AsImplementedInterfaces());
```

In the Type method, you give the library some marker/representative type, and it then grabs the `Assembly` and goes from there.

Broadly, I'd rather have some way to reference an assembly without having to go through some type first. That's a bit more to do, something like `assemblyof(System.Core)` vs `assemblyof(string)` or `typeof(string).Assembly`.

**Reply by jzabroski** [reply]
Why would you pass in a `Type` rather than a fully qualified assembly name? Note, assembly names don't map to any namespace, they map to assembly names. They're functionally ".NET URN Prefixes".

See my thoughts here: https://github.com/dotnet/csharplang/discussions/6229#discussioncomment-2986041

Note, in general, I don't see why the user of `assemblyof` should have to know what types are in that assembly. The whole point is to provide a succinct universal quantifier/existential quantifier that checks an assembly and does a consequent action.

**Comment by HaloFour** [comment]
It sounds like that use case already has a perfectly good solution of offering overloads that let you specify a `Type` to which that assembly belongs.  And even if it didn't, `typeof(T).Assembly` is readily available.  I'd like to hear of situations where a feature like this would save more than single-digit keystrokes across an entire project.

**Comment by marxxxx** [comment]
readability would be another advantage of this operator. picking an arbitrary type to just get to the assembly feels like a workaround and makes the reader think: "is there something special to this type i just don't get?"

**Comment by theunrepentantgeek** [comment]
@marxxxx - you'd still have to pick an arbitrary type to use with **typeof**. @jbogard even said as much.

There's nothing else in the C# language that could be used as a replacement. The obvious candidate, is namespaces, but those can be spread across multiple assemblies.

**Comment by Richiban** [comment]
I also don't see what this gives you that you can't get with `typeof(T).Assembly`, but I would like this feature if it meant you had strongly-typed access to an assembly _without_ having to pick a random type from that assembly.

So, rather than 

```csharp
var assembly = typeof(JObject).Assembly;
```

we could write

```
var assembly = assemblyof(Newtonsoft.Json);
```

This has the slight advantage of making it easier (possible?) to reference an assembly, in a strongly-typed way, that might have no public types (does happen sometimes).

**Comment by dsaf** [comment]
What if `memberof` https://github.com/dotnet/csharplang/issues/712 on a namespace gives the assembly?

**Comment by HaloFour** [comment]
@Richiban 

> This has the slight advantage of making it easier (possible?) to reference an assembly, in a strongly-typed way, that might have no public types (does happen sometimes).

I kind of doubt it.  IIRC there's no IL opcode to directly reference an assembly like there is for types (`ldtoken`), and the easiest way to get an Assembly instance is via the BCL is through `Type.Module.Assembly` (which is what `Assembly.GetAssembly` does).

There's also the language question of how to reference the assembly without the use of types.  Assembly names don't have to follow the rules of dotted identifiers.  They're only limited to the rules of what characters may be used in file names in Windows.

@dsaf 

Namespaces can span many assemblies.  They don't exist beyond a component of the type names within the assembly.

**Comment by yaakov-h** [comment]
Assembly names also don’t have to match file names, and you can reference multiple assemblies with the same name and different versions IIRC...

**Comment by svick** [comment]
@jbogard

> Broadly, I'd rather have some way to reference an assembly without having to go through some type first. That's a bit more to do, something like `assemblyof(System.Core)` vs `assemblyof(string)` or `typeof(string).Assembly`.

How would you deal with the issue that runtime assemblies and reference assemblies can disagree? For example, `typeof(string).Assembly` returns the System.Private.CoreLib assembly on .Net Core, but the compiler only sees that `System.String` comes from the reference assembly System.Runtime.

When using reflection, you probably want the System.Private.CoreLib assembly, but the compiler has no knowledge about it, so I don't think it could allow `assemblyof(System.Private.CoreLib)`.

**Comment by tobyee** [comment]
Looks like `#define` in C++, create a macro to save characters and improve readability .

**Proposal 1**: Create a new preprocessor directive(or may be share the old one #define) in C# like this 
```cs
#macro assemblyof(T) (typeof(T).Assembly)
```
**Proposal 2**: Static using which was already supported but could be make a little change.
```cs
//foo.cs
public MyCustomOperaters {
  public static Assembly assemblyof<T>() { return typeof(T).Assembly; }
}
//use.cs
using static MyCustomOperaters;
...
assemblyof<aaa>(); // current syntax, I think 2 characters more is acceptable.
assemblyof<aaa>;  //  new syntax: brackets can be ignore when the generic method has no parameters
...
```

**Comment by YairHalberstadt** [comment]
```csharp
assemblyof<aaa>;  //  new syntax: brackets can be ignore when the generic method has no parameters
```

This already has a meaning:

```csharp
Action M() => () => Console.WriteLine("Hello World");

...

Action m = M()
m(); //prints "Hello World"

...

Action m = M;
m(); //does nothing

**Comment by yaakov-h** [comment]
@YairHalberstadt 

> error CS0407: 'Action M()' has the wrong return type

**Comment by YairHalberstadt** [comment]
@yaakov-h

Ah well. If I had a computer on me I could find an example where it would be ambiguous.

**Comment by tobyee** [comment]
@YairHalberstadt 
You means generic class？
```cs
namespace Something {
    public static class AssemblyExtensions  {
        public static Assembly assemblyof<T>() { return typeof(T).Assembly; }
    }
    public class assemblyof<T> {
        public static string FullName { get; }
    }
}

namespace Something.Test {
    using static AssemblyExtensions;
    public class Foo {
        public void Test() {
            var c = assemblyof<int>.FullName;       // ambiguous???
        }
    }
}
```

**Comment by YairHalberstadt** [comment]
@tobyee
Nope. I mean that a method without a following paranthesis is a way of creating a delegate from that method.
